feat(context-providers) Enable arguments to be passed as property object

moofish32 opened the pull request:
This is a first draft at updates originally from #425

I am attempting to apply some types to context providers where there were just arrays. 
The arrays were good because of ordering, but I've attempted to preserve that with a sort. 

I would like to implement a VPC Provider because I find imports for VPCs painful. In addition
to solving a customer problems VPC Provider will stress the ability to return a complex type.

Is this design in the right direction?

TODO - fix commit message this is a BREAKING Change

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

moofish32 commented:
I'll go back and remove the package lock diffs, my apologies.

moofish32 commented:
I need a hint on where this integ test failure is located. 

```
@aws-cdk/aws-autoscaling: A reserved context key ('context.aws:') key was found in /home/travis/build/awslabs/aws-cdk/packages/@aws-cdk/aws-autoscaling/test/cdk.json, it might cause surprising behavior and should be removed.
@aws-cdk/aws-autoscaling: A reserved context key ('context.aws:') key was found in /home/travis/build/awslabs/aws-cdk/packages/@aws-cdk/aws-autoscaling/test/cdk.json, it might cause surprising behavior and should be removed.
@aws-cdk/aws-autoscaling: Need to perform AWS calls for account 12345678, but no credentials found. Tried: default credentials.
@aws-cdk/aws-autoscaling: Error: Command exited with status 1
@aws-cdk/aws-autoscaling:     at exec (/home/travis/build/awslabs/aws-cdk/tools/cdk-integ-tools/lib/integ-helpers.js:102:15)
@aws-cdk/aws-autoscaling:     at IntegrationTest.invoke (/home/travis/build/awslabs/aws-cdk/tools/cdk-integ-tools/lib/integ-helpers.js:50:20)
@aws-cdk/aws-autoscaling:     at <anonymous>
@aws-cdk/aws-autoscaling: Error: cdk-integ-assert exited with error code 1
```
I can't seem to trace this back through the integ helper.

moofish32 commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -306,10 +312,10 @@ async function getCLICompatibleDefaultRegion(profile: string | undefined): Promi
 
   let region = process.env.AWS_REGION || process.env.AMAZON_REGION ||
-      process.env.AWS_DEFAULT_REGION || process.env.AMAZON_DEFAULT_REGION;
+  process.env.AWS_DEFAULT_REGION || process.env.AMAZON_DEFAULT_REGION;
```
not sure why this is autoformatting this way, but I'll fix this.

moofish32 commented on packages/@aws-cdk/cx-api/lib/cxapi.ts:
```diff
@@ -24,10 +24,17 @@ export type CXRequest = ListStacksRequest | SynthesizeRequest;
  * (should have been an interface, but jsii still doesn't have support for structs).
  */
-export interface MissingContext {
+export interface ContextProviderProps {
```
after reading this again (a few days later). I think I'm ready to switch this back to `MissingContext`.

```ts
export interface MissingContext {
  provider: string;
  props: {
    account: string;
    region: string;
    [key: string]: string;
  };
}
```

moofish32 commented:
@rix0rrr @eladb looking for a little feedback on the direction here.

My end goal is to enable a significantly improved capability for VPC Import. Right now you have to know the `vpcIds` and all the `subnetIds` in order to use a `VpcPlacementStrategy`. 

Instead of having do something like

```
export class MyStack extends cdk.Stack {
    constructor(parent: cdk.App, name: string, props?: cdk.StackProps) {
        super(parent, name, props);
        const vpc: ec2.VpcNetworkRef = ec2.VpcNetwork.import(this, 'MyVPC', {
            vpcId: new ec2.VPCId(props.vpcId),
           availabilityZones: props.availabilityZones,
           publicSubnetIds: props.subnetIds,
           isolatedSubnetIds: [],
        });
       // create VpcPlacementStrategy
       // create my asg or other resources (e.g. lambda, eks, ecs etc)
       // I would rather 
      const vpc: ec2.VpcNetworkRef = ec2.VpcProvider(this, { vpcId: 'vpc-xxxxxxxx' });
      // or
      const vpc2: ec2.VpcNetworkRef = ec2.VpcProvider(this, { cidrBlock: '192.168.0.0/16'});
      // essentially enable the filters you see here: 
      // https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-vpcs.html
   }
}
```
In addition to Vpc we would also enable the AMI look up feature that would not rely on the ssm parameter method we have now. Obviously the type checking and casting behavior would all be hidden from the user, but we have to use some `any` values to enable this code behind the scenes.

Most important question I have is: Is this a good feature to spend time implementing and am I headed in the right direction?

eladb commented:
@rix0rrr  can you please take a look at this?

rix0rrr commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -16,39 +18,56 @@ const SSM_PARAMETER_PROVIDER = 'ssm';
+  private readonly props: {[key: string]: any};
 
-  constructor(private context: Construct) {
+  constructor(private context: Construct, provider: string, props: {[key: string]: any} = {}) {
```
If you declare the arguments as:

```
constructor(..., private readonly provider: string, private readonly props: {[key: string]: any} = {}) {
}
```

You don't have to declare the member variables above and don't have to assign them either, [that happens automatically](https://kendaleiv.com/typescript-constructor-assignment-public-and-private-keywords/).

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -1,10 +1,10 @@
 
 export interface ContextProviderPlugin {
-  getValue(scope: string[], args: string[]): Promise<any>;
+  getValue(filter: cxapi.ContextProviderProps): Promise<any>;
```
"filter"?

moofish32 replied:
how about `provider: cxapi.MissingContext` instead?

rix0rrr replied:
After moving `account` and `region` into the dictdionary, is there a reason to pass the entire `ContextProviderProps` object anymore?

We can just do:

```ts
getValue(args: {[key: string]: any})
```
right?

moofish32 replied:
ahh - a little different than what I first thought, but let me try that.

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
+    const props: cxapi.HostedZoneProviderProps = filter.props;
+    const domainName = props.domainName;
+    const privateZone: boolean = !!props.privateZone;
+    const vpcId =
```
??

moofish32 replied:
yeah -- this looks like a coding on a plane bonus bug ✈️ 🐛 

Should we expand the scope of this PR and find a away to integration test these?

rix0rrr replied:
I'm not opposed to it, but I have no clever ideas on how to do it that wouldn't take a long time to build and/or make tests complicated to run.

For example, if we're going to test the actual service call, we'd need AWS credentials and network access on every build--something we can (mostly) avoid now.

You might potentially consider mocking/stubbing the external APIs, OR having a test that you run manually, and a signature (like a hash of the file) to assert that you manually ran the test on the latest version of the source.

moofish32 replied:
I haven't fully looked at this yet, but perhaps you can tell me I'm walking into a big thing.

My thought was to keep this as simple as possible. Enable cdk-integ to be able to run more than one stack. Create resources in the first stack and use providers to find those resources in the second stack. Then sanitize the cdk.json file with the appropriate region/account information such that AWS API credentials are not required after the developer runs the integration test locally.

This definitely isn't simple, but would that work and not be a nightmare to implement?

rix0rrr replied:
It's definitely a bit out of the mold of our current integ tests.

* You can't comfortable write this test in `aws-cdk` package (I don't think), it has to go in one of the library packages. Which one?
* How are we going to specify the order between multiple stacks in `cdk-integ`? We have explicitly punted on that because there's no way to automatically infer it, and no place to specify it.
* We don't check in `cdk.json` usually, and in fact `cdk-integ` automatically deletes it. To be able to run an offline test we should keep that file; but preferably not all files in all cases, so only *some* instances of that file...

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
+    return candidateZones[0];
+  }
+
+  private filterZones(r53: AWS.Route53, zones: AWS.Route53.HostedZone[], privateZone: boolean, vpcId: string | undefined): AWS.Route53.HostedZone[] {
```
async

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
+      if (vpcId) {
+        const vpcZones: AWS.Route53.HostedZone[] = [];
+        for (const zone of candidates) {
+          r53.getHostedZone({Id: zone.Id}, (err, data) => {
```
This is not going to do what you want, because it's going to be asynchronous.

Just add an `await`.

moofish32 commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -60,50 +79,65 @@ export class ContextProvider {
+      switch (typeof obj[key]) {
+        case 'object': {
+          const childObjStrs = this.objectToString(obj[key]);
+          const qualifiedChildStr = childObjStrs.map( child => (`${key}${child}`)).join(':');
```
Ok - so I started by blocking this. However, I have found it very useful to allow tags to be used in providers (from other frameworks that provide similar function). Both AMIs and VPCs are great examples where tags will be useful. So I would prefer to add the separator and keep the use case.

rix0rrr replied:
Cool.

rix0rrr commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -16,39 +18,56 @@ const SSM_PARAMETER_PROVIDER = 'ssm';
   }
 
+  public get key(): string {
+    const account = this.account;
```
Yes me too.

moofish32 commented:
Ok moving this thread to a comment https://github.com/awslabs/aws-cdk/pull/823/files/f6c1ec2ffd29fec8909b90cb7daaf311013ab610#r223964075 because github hides it. 

I understand what you are saying. I did think the integ tests would need a new type `provider.ssm.ts` instead of `integ.<L2>.ts` in the test folder. However, looking at cdk-integ this is bigger than I can do in this PR. The other question is does the team think this is worth it?

Should we move the providers into the `@aws-cdk` packages in this?
AvailabilityZone -> `@aws-cdk/aws-ec2`
SSM -> `@aws-cdk/aws-ssm`
HostedZone -> `@aws-cdk/aws-route53`

I'll need to do some more manual testing as well.

rix0rrr commented:
> Should we move the providers into the @aws-cdk packages in this?
No, they're definitely part of the toolkit.

For now, we don't really have a better answer than manual testing. This is admittedly not great and we're aware of this and thinking about it, but this is the state of the world.

rix0rrr commented on packages/@aws-cdk/cx-api/lib/cxapi.ts:
```diff
@@ -26,8 +26,17 @@ export type CXRequest = ListStacksRequest | SynthesizeRequest;
-  scope: string[];
-  args: string[];
+  props: {
+    account: string;
```
I believe `account` and `region` are still potentially `undefined` (for providers that don't require them--even though we don't have any right now), so add a `?` to them.

rix0rrr commented on packages/@aws-cdk/cx-api/lib/cxapi.ts:
```diff
@@ -26,8 +26,17 @@ export type CXRequest = ListStacksRequest | SynthesizeRequest;
+  };
+}
+
+export interface HostedZoneProviderProps {
```
I feel this should be defined next to the `HostedZoneProvider` class. 

I don't feel comfortable using one struct for both the API properties and defining the wire protocol, because they're distinct things. It's probably okay to have an implicit protocol on the wire (or define 2 structs for 2 distinct purposes).

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
+      if (vpcId) {
+        const vpcZones: AWS.Route53.HostedZone[] = [];
+        for (const zone of candidates) {
+          await r53.getHostedZone({Id: zone.Id}, (err, data) => {
```
I should have been more specific here. I'm not sure whether this works or not, but it sure looks weird to me. I actually meant doing the following:

```ts
const data = await r53.getHostedZone({ Id: zone. Id }).promise();
// No need to check for err, just look at data afterwards
```

It's the same pattern as for the `listHostedZonesByName` call you're doing in the other function, and basically the way all SDK calls should look.

The `.promise()` at the end is a weird convention that the SDK has in there (they could have made it so it works without it, but they didn't). It makes it possible to `await` this API call.

rix0rrr commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
+        for (const zone of candidates) {
+          await r53.getHostedZone({Id: zone.Id}, (err, data) => {
+            if (err) {
+              throw new Error(err.message);
```
Not that it matters because you're going to be rewriting this call sequyence anyway, but throwing an `Error` in an async callback is almost never what you want to be doing.

In an async callback your call stack is gone: you're basically a couple of stack frames deep in an event loop, and throwing an error is basically only going to abort your callback and the rest of the failure is lost. To pull the thread all the way through, you'll generally call another callback with the error (boooh) or report the error via a promise (yay).

But the `await` syntax takes care of all of that for you.

rix0rrr commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -17,38 +19,48 @@ export class ContextProvider {
+    private readonly context: Construct,
+    private readonly provider: string,
+    private readonly props: {[key: string]: any} = {}) {
     this.stack = Stack.find(context);
```
I think we just go like this in the constructor:

```
this.props.account = this.stack.env.account;
this.props.region = this.stack.env.region;
```

And then in all other places we just use `this.props`.

rix0rrr commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -17,38 +19,48 @@ export class ContextProvider {
     // return the default value
-    if (!scope) {
-      this.context.addError(formatMissingScopeError(provider, args));
+    if (!this.account || !this.region) {
```
Interestingly, this now means something else. Because previously there was a distinction between "not having a scope" and "not having an account/region", and now there isn't anymore.

This only comes into play if we have context providers that don't need both an account and a region, which we don't have right now anyway so I'm okay with this change and we can always fix it once we do. But it's something to consider.

moofish32 commented:
@rix0rrr  -- I used this to test the provider:

```
#!/usr/bin/env node
import r53 = require('@aws-cdk/aws-route53');
import ec2 = require('@aws-cdk/aws-ec2');
import cdk = require('@aws-cdk/cdk');

class TestLatestCdkStack extends cdk.Stack {
  constructor(parent: cdk.App, name: string, props?: cdk.StackProps) {
    super(parent, name, props);
    const vpc = new ec2.VpcNetwork(this, 'MyVpc', {
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'Public',
          subnetType: ec2.SubnetType.Public,
        }
      ]
    });

    new r53.PrivateHostedZone(this, 'PrivateZone', {
      zoneName: 'vpc.internal',
      vpc,
    });

    new r53.PublicHostedZone(this, 'PublicZone', {
      zoneName: 'tdfa.xyz',
      comment: 'CDK created',
    });

  }
}

class TestLatestHosted extends cdk.Stack {
  constructor(parent: cdk.App, name: string, props?: cdk.StackProps) {
    super(parent, name, props);
    new r53.PublicHostedZone(this, 'PublicZone', {
      zoneName: 'tdfi.xyz',
      comment: 'CDK created',
    });
    const hostedZoneProps: r53.HostedZoneRefProps = new r53.HostedZoneProvider( this, { domainName: 'tdfa.xyz', }).findHostedZone();
    const zone = r53.HostedZoneRef.import(this, 'Imported', hostedZoneProps);
    new cdk.Output(this, 'Test', {
      value: zone.hostedZoneId,
    });
    const privateZoneProps: r53.HostedZoneRefProps = new r53.HostedZoneProvider( this, {
      domainName: 'vpc.internal',
      privateZone: true,
      vpcId: 'vpc-093faadfc4f1ba5df',
    }).findHostedZone();
    const privateZone = r53.HostedZoneRef.import(this, 'ImportedPrivate', privateZoneProps);
    new r53.TXTRecord(zone, 'txt', {
      recordName: 'mytxt',
      recordValue: 'special text message for validation',
    });
    new r53.TXTRecord(privateZone, 'privateTxt', {
      recordName: 'mytxt',
      recordValue: 'special text message for validation',
    });
  }

}
const app = new cdk.App();

new TestLatestHosted(app, 'TestHosted');
new TestLatestCdkStack(app, 'TestLatestCdkStack');

app.run();
```

I moved the provider (non AWS SDK side -- do we have a name for the split?) into the route53 module because I wanted to return the data structure for `r53.HostedZoneRef.import( ... )`. Sorry to make some changes this late in the process (specifically I've changed the function name to `findHostedZone`). I'm not sure if we should make this a generic name or a specific name per provider?

moofish32 commented on packages/@aws-cdk/cdk/lib/context.ts:
```diff
@@ -16,95 +22,108 @@ const SSM_PARAMETER_PROVIDER = 'ssm';
-    if (!stack) {
-      throw new Error(`${providerDescription}: construct must be in a stack`);
-    }
+  public getStringListValue(
```
I think I can dry this function up by passing a function here. I don't think I can use generics here. I suppose I could extract this to a non exported function and pass everything into it, but I think just passing a `validateType(value: any): boolean{}` will probably do?

moofish32 replied:
actually that is also blocked by JSII ... I'm not sure we can really DRY this up much then?

rix0rrr replied:
It's fine as-is.

rix0rrr commented:
Looks good! Final touches and then we can get it in.

moofish32 commented on packages/@aws-cdk/aws-route53/lib/hosted-zone-provider.ts:
```diff
@@ -0,0 +1,57 @@
+  /**
+   * Return the hosted zone meeting the filter
+   */
+  public findHostedZone(): HostedZoneRefProps {
```
Great idea 👍 I will add.

moofish32 commented:
@rix0rrr -- I believe I have implemented the final changes, trying to fix that conflict with the last push. Anything else you see from me? I believe this is breaking, but open to trying to make this more backwards compatible if we need to.

moofish32 commented on packages/aws-cdk/lib/contextplugins.ts:
```diff
@@ -48,21 +53,89 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
   }
 }
 
+export interface HostedZoneProviderProps {
```
Docs here
